- hyperchain/contract-pubkeys endpoint

klercker opened the pull request:
- endpoint implemented
- method to get contract pubkeys from HC consensus module (which also considers the staking contract - change from before
- updated HC test suite to configure all contracts (rewards == staking)

- ATM endpoint fails completely if any of the contracts are not configured. The consensus modeul probably shouldn't even start if this is the case. Right now this is actually due to the ETS cache failing if the contracts aren't there, so there needs to be some fixing around there to handle this in a good way if allowed.


This PR is supported by the Aeternity foundation

ThomasArts commented on apps/aecore/src/aec_consensus_hc.erl:
```diff
@@ -474,6 +475,16 @@ rewards_contract_pubkey() ->
+                  _ -> erlang:error({contract_owner_not_valid_contract, EncContractId})
+              end
+          end,
+    aeu_ets_cache:get(?ETS_CACHE_TABLE, rewards_contract_pubkey, Fun).
```
Why here the rewards contract pubkey as default???

klercker replied:
CopyPaste error, will fix. ran since rewards and staking are the same in the current setup.

ThomasArts commented on apps/aehttp/src/aehttp_dispatch_ext.erl:
```diff
@@ -909,6 +910,14 @@ handle_request_('GetPinningTx', _Params, _Context) ->
+handle_request_('GetHyperchainContractPubkeys', _Params, _Context) ->
+    % TODO handle not in HC at all or consensus not initialized?
+    lager:debug("HyperchainsGetContract"),
+    {200, [], #{<<"staking">> => aeser_api_encoder:encode(contract_pubkey, aec_consensus_hc:get_contract_pubkey(staking)),
```
Here we spill over the logic on how many contracts there are and what they are named from consensus to the API. The names  `election`, `rewards` are macros in the consensus code, but now hard-coded in the API.

Not sure about this, but my feeling is that it would be cleaner to ask the consensus module for all contracts and have that module return this map. In that way, the logic of how many and what contracts we have can be adapted rather easily.... but the price is that we cannot say in the API specification what we expect back other than an object with keys and contract addresses.

klercker replied:
Sure, can do that no prob. Not sure how the endpoint will feel about getting a map with unspeced properties though. What's the yaml construct for that?

ThomasArts reviewed (approved):
Might consider not hard-coding naming of contracts. On the other hand, if SDK is to use this, they need to know a bit on what kind of contracts they can expect.

ThomasArts commented:
Failure on otp26 is a state channel test. Seems to me unrelated to code delivered. Please have a look and run again.

ThomasArts reviewed (commented):
- [ ] Look at reward_contract as default!
- [ ] rebase on master

klercker commented:
Fixed the cache key and rebased. I say we merge, and if we decide to go with the less hardcoded route (map of pubkeys) we can implement that then? But you guys decide.

hanssv commented on apps/aehttp/test/aehttp_hyperchains_SUITE.erl:
```diff
@@ -818,6 +820,33 @@ epochs_with_fast_parent(Config) ->
+    StakingContractPK = rpc(Node, aec_consensus_hc, get_contract_pubkey, [staking]),
+    ElectionContractPK = rpc(Node, aec_consensus_hc, get_contract_pubkey, [election]),
+    RewardsContractPK = rpc(Node, aec_consensus_hc, get_contract_pubkey, [rewards]),
+    %% note: the pins are for the last block in previous epoch
```
I'm not entirely sure what this has to do with the code...

klercker replied:
I test what I get back from the call is identical to what was in the contract?

klercker replied:
But good Q about that comment. C&P leftovers I presume. Gone now, will be in my next functional commit.

hanssv replied:
The comment is still there, not resolved.

klercker commented:
@hanssv are we having timing issues on OTP 24? A test for block type in `hc.verify_fees` fails on 24 but is ok on 25/26?

klercker commented:
@hanssv @ThomasArts any suggestions what we do to get pass this? Have re-run 3 times with no luck. same `hc.verify_fees` block type check (in `produce_cc_to_height`). The failure seems to be the first time blocks are produced in the first epoch. But it's beyond me why it would fail in 24. Is there a way to run ct-latest on OTP24 locally? Can't find it in Makefile at least, but perhaps rebar has options?

hanssv commented:
> @hanssv @ThomasArts any suggestions what we do to get pass this? Have re-run 3 times with no luck. same `hc.verify_fees` block type check (in `produce_cc_to_height`). The failure seems to be the first time blocks are produced in the first epoch. But it's beyond me why it would fail in 24. Is there a way to run ct-latest on OTP24 locally? Can't find it in Makefile at least, but perhaps rebar has options?

Running on OTP24 should be a breeze with devcontainers and VSCode, right? I've got all OTP versions 22-27 locally and switch between them when I need to. Switching OTP-version can't be done in the Makefile.

I think the test fails because it doesn't get a micro-block when expected, the reason is that the test falls behind, and when the Transaction is submitted the producing node is already past the micro-block producing point. Possible solutions are make the test more robust in this case, or, make block-time longer (thus making it less likely to fall behind, but also slower!).

klercker commented:
Tests run fully!! If everybody is reasonably happy with the state of things (We are at tip of master, I rebased this morning based on @ThomasArts merge from yesterday), I suggest approving and getting this into the tree now :)

hanssv commented on apps/aehttp/priv/oas3.yaml:
```diff
@@ -535,6 +535,27 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/Error"
+  /hyperchain/contract-pubkeys:
```
Why the obsession with the term 'pubkey' here? What else would the endpoint return (the bytecode?) I'd name the endpoint `/hyperchain/contracts`?

klercker replied:
True. Will resolve these two then. Look forward to new CCI run and re-approval...
